### PR TITLE
Missing block headers

### PR DESCRIPTION
**Pull request by adamsmo**
This PR handle case when we have requested block bodies by hashes for block headers that are not in our DB, this may happen because of killing client while persisting data

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 1]
```diff
@@ -276,14 +281,24 @@ trait FastSync {
...
-        case (Some(header), body) =>
-          val result = validators.blockValidator.validateHeaderAndBody(header, body)
-          result.isRight
-        case _ => false
-      }
+    private def validateBlocks(requestedHashes: Seq[ByteString], blockBodies: Seq[BlockBody]): ValidationResult = {
+      var result: ValidationResult = Valid
+      (requestedHashes zip blockBodies)
+        .map { case (hash, body) => (blockchain.getBlockHeaderByHash(hash), body) }
+        .forall {
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: to stop on first error

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 2]
```diff
@@ -473,6 +488,11 @@ object FastSync {
   private case object PersistSyncState
   case class MarkPeerBlockchainOnly(peer: Peer)
 
+  private sealed trait ValidationResult
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: maybe BlockBodyValidationResult?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 3]
```diff
@@ -276,14 +281,24 @@ trait FastSync {
...
+      var result: ValidationResult = Valid
+      (requestedHashes zip blockBodies)
+        .map { case (hash, body) => (blockchain.getBlockHeaderByHash(hash), body) }
+        .forall {
+          case (Some(header), body) =>
+            val validationResult: Either[BlockValidator.BlockError, Block] = validators.blockValidator.validateHeaderAndBody(header, body)
+            result = validationResult.fold(_ => Invalid, _ => Valid)
+            validationResult.isRight
+          case _ =>
+            blockBodiesQueue = Seq.empty
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: can we move this side effects out of this function?

Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: actually I'm not sure if it's good idea to clear the queues completely, maybe we can only remove the bodies/receipts for the headers we don't have?

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: side effects - good idea
filtering failure - I think clearing all is simpler than checking which block header are missing and deciding base on that, also i we clear we do not have to check every block hash for missing block header - if we find first we fail

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 4]
```diff
@@ -226,6 +222,22 @@ trait FastSync {
...
+      validateBlocks(requestedHashes, blockBodies) match {
+        case Valid =>
+          insertBlocks(requestedHashes, blockBodies)
+        case Invalid =>
+          blacklist(peer.id, blacklistDuration, s"responded with block bodies not matching block headers, blacklisting for $blacklistDuration")
+          self ! FastSync.EnqueueBlockBodies(requestedHashes)
+        case DbError =>
+          blockBodiesQueue = Seq.empty
+          receiptsQueue = Seq.empty
+          bestBlockHeaderNumber = bestBlockHeaderNumber - 2 * blockHeadersPerRequest
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: what kind of magic formula is this? :P
maybe we can just set it to the first missing header num?

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: just rough estimate of safety buffer ;) , but yes we can go with first missing header

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: (And then it won't need precedence brackets for readability)

**Comment by adamsmo** [comment]
@mcsherrylabs  will add TODO

**Pull request merged**